### PR TITLE
Fix goenv plugin

### PR DESCRIPTION
**Pull request by davidpfarrell**
The original version suffers from a few bugs:

* If the user defined `$GOENV_ROOT`, **but** defined it outside of `$HOME/.goenv`, then the script would exit without finding it.
* For path-munging, the script checked for the tool in `$HOME/.goenv` instead of `$GOENV_ROOT` - This would cause the tool to **not** end up on the path if the `$GOENV_ROOT` was defined outside of `$HOME`
* The script would path-munge `$GOENV_ROOT/bin` even if the binary was already on the path (say by using homebrew), causing it now to appear on the path twice, with the `$GOENV_ROOT` version overriding the homebrew version.

### CODE REVIEW

**Comment by cornfeedhobo** [comment]
Hey. I owe you an apology. I could have used a custom plugin entry and kept up my progress locally while we discussed everything. I'm sorry.

Would you be willing to expand this to all the env plugins? Each one has minor differences, so you'll want to do some reading on each one, but they're similar enough. I think the consistency will really help. Thanks.

**Comment by cornfeedhobo** [comment]
Before I get started, I hope I can clear the air and say I'm sorry our last exchange got so heated.

Would you please expand this to all the env plugins? As we discussed in the other PRs, the existing state is a pattern that other env plugins are using. I think you've made some great points and catches, but I'd rather not pick arbitrary places to enforce this level of scrutiny and think the consistency will really help everyone, which is the point of this project.

Each *env tool has minor differences, so you'll want to do some reading on each one, but they're similar enough.

Thanks!

**Comment by davidpfarrell** [comment]
Hello @cornfeedhobo ,

> I hope I can clear the air and say I'm sorry our last exchange got so heated.

Thank you very much for taking the time and effort to apologize - I also hope we can clear the air.  I suspect we're going to have a lot of conversations over time as we're both active contributors who care a lot about the BASH-IT project.  But I'm also aware that we have very different view points of how to do things, and that's going to be a challenge.  I hope we can both avoid letting discussions get too heated and instead focus on keeping BASH-IT moving forward.

> Would you please expand this to all the env plugins?

Two of the mantras I use as part of working to enable my teams to move fast while keeping code quality high are:
* `"Keep changes tight and focused"`
* `"Minimize the size of the diff"`

These mantras focus on minimizing risk and keeping change-sets singularly-focused and easy to reason about.

Trying to modify core logic of multiple plugins in one PR adds too much complexity and risk.

So, while I believe there is plenty of room for improvement across the existing *env plugins, I would not be interested in expanding this PR to encompass multiple plugins.

What I would be **very interested** in doing, however, is having a discussion on formalizing the expected behavior of these types of plugins (and maybe all plugins in general).

If *you* wanted to open an issue to discuss formalizing + back-fixing *env plugin behavior, I would absolutely participate in the discussion.

Once we reached consensus (me, you, @nwinkler, etc), we could update the documentation (ie. maybe `CONTRIBUTING.md` etc), identify each plugin that should be subject to the new requirements, and then open issues to discuss/address each plugin *individually*.

Once we get to that point, I'll be happy to pick up some of the *env plugins and open PRs to fix them.

> As we discussed in the other PRs, the existing state is a pattern that other env plugins are using.

I put extra effort in this PR to ensure that it matches the *spirit* and *patterns* you established in your original PR.  I hope that comes through as you review the changes.

This PR is not focused on trying to change any patterns, but is instead focused on fixing *actual* bugs, the most significant being better support for users who have configured `$GOENV_ROOT` to point outside of `$HOME/.goenv`

Much of the crux of the bug stems from this line:
```
export GOENV_ROOT="${GOENV_ROOT:-$HOME/.goenv}"
```

This line's `conditional-assign` logic signifies that we care about the user's ability to specify the location of the root folder.

In order to fully honor the user's ability to control this value, we need to also check there for the binary before deciding to bail.

And although neither the `pyenv`, `rbenv`, or `jenv` plugins follow this behavior, I'm very much in support of the user's ability to control this value, so I would not advocate for removing this logic, even though it goes against the "`existing pattern`".

I hope this clarifies the intention of the PR, and also hope that we can get these bug-fixes merged quickly.

Thank you for your time, and please feel free to \@mention me if you open an issue to discuss formalizing the *env plugin behavior.

-DF

**Comment by nwinkler** [comment]
Thanks for the good discussion on this topic. I'm with @davidpfarrell on this, let's keep the changes localized as much as possible.

Enforcing the same conventions across the whole codebase is a nice goal, but is going to be rather difficult. I'm much more in favor of fixing a current issue now than waiting for a much bigger PR (which will be more work to review)...

I'm in favor of merging this PR - how do you feel about that, @cornfeedhobo?

**Comment by cornfeedhobo** [comment]
Unfortunately I've been peeling away too much time to respond in as much detail as I would like, so please excuse my terse reply. I've watched a few companies fall apart because of these maxims. I'm actually at a large enterprise trying to undo these habits and align development teams. It's funny that we are discussing this around the `goenv` plugin because Rob Pike has been a huge figure in pushing away from these maxims.

By not updating all conventions at once, you leave a path for future contributors that is confusing and more difficult to navigate. They may look at `rbenv`, make a new plugin, only to get a **wall of text** and stop there. The real cost of programming is human hours and stress, and this maxim makes that cost even higher.

I know you are senior engineers. I'm more than confident that a PR encompassing all the env plugins would be trivial to review.

P.S. personally, I kinda still want `GOENV_ROOT` blindly exported without checking for the tool, but I don't care enough to make a thing of it.

**Comment by cornfeedhobo** [comment]
@nwinkler  Unfortunately I've been peeling away too much time to respond in as much detail as I would like, so please excuse my terse reply.

I've watched a few companies fall apart because of these maxims. I'm now at a large enterprise, where my position is specifically to undo these habits and align development teams. It's funny that we are discussing this around the `goenv` plugin because Rob Pike has been a huge figure in pushing away from these maxims. I encourage every senior engineer to watch his talks on efficiency and collaboration.

By not updating all conventions at once, you leave a path for future contributors that is confusing and more difficult to navigate. They may look at `rbenv`, make a new plugin, only to get a review that is a **wall of text** and stop there. The real cost of programming is human hours and stress, and this maxim makes that cost even higher. Look at the backlog of PRs and open Issues; the easier we can make things for the average contributor, all the better.

I know you are both senior engineers. I'm more than confident that a PR encompassing all the env plugins would be trivial to review, or you can break them up as a suite of PRs if you want clean reverts, but I still think they should be done in batch.

Personally, I still like `*_ROOT` blindly exported without checking for the tool, but I don't care enough to make a thing of it.

P.S. On a personal note, having my work taken and a new branch started was like a punch to the stomach. I'm trying to press forward, but it still really bugs me. I've been contributing to codebases since I was around 12. I had never had anyone take my code like that. I feel robbed and belittled. Frankly I'm not sure I want to continue if it's going to be like this going forward.

**Comment by cornfeedhobo** [comment]
@nwinkler  Unfortunately I've been peeling away too much time to respond in as much detail as I would like, so please excuse my terse reply.

TL;DR: yeah it's fine, merge it

My real opinion: I've watched a few companies fall apart because of these maxims. I'm now at a large enterprise, where my position is specifically to undo these habits and align development teams. It's funny that we are discussing this around the `goenv` plugin because Rob Pike has been a huge figure in pushing away from these maxims. I encourage every senior engineer to watch his talks on efficiency and collaboration.

By not updating all conventions at once, you leave a path for future contributors that is confusing and more difficult to navigate. They may look at `rbenv`, make a new plugin, only to get a review that is a **wall of text** and stop there. The real cost of programming is human hours and stress, and this maxim makes that cost even higher. Look at the backlog of PRs and open Issues; the easier we can make things for the average contributor, all the better.

I know you are both senior engineers. I'm more than confident that a PR encompassing all the env plugins would be trivial to review, or you can break them up as a suite of PRs if you want clean reverts, but I still think they should be done in batch.

**Comment by nwinkler** [comment]
@cornfeedhobo Thanks for explanation - good thoughts.

I fully agree with you (and I think that @davidpfarrell would also agree), I'm fully in line with cleaning things up when the opportunity is there. There are a couple of things that are different here, though:

Bash-it is an open source project, it's not a company. At work, I push my teams to complete refactorings and clean-up work as part of the regular work - don't delay these things. With open source work, it's slightly different. We rely on voluntary contributions, which people mostly do in their free time. I can ask contributors to include other works in their Pull Requests, but if I push too hard, people will shy away.

Some other examples were people were asking for bigger changes are putting an open source license on Bash-it, or providing error feedback to the user in case of a missing dependency. These are all good things to have, but the fact is that these are going to be a lot of work - and the reality is that the majority of contributors don't have the time to implement a big change like that.

Both you and @davidpfarrell are some of the more active (probably the most active at the moment) contributors - I really appreciate this. This might give us an opportunity to do some bigger work if we plan carefully. I'd love to get some of this stuff done, but not sure how we want to go about it.

One option might be to create a new _project_ here in Bash-it (https://github.com/Bash-it/bash-it/projects) for these kind of refactorings, and start listing items there (like the Pyenv plugin that you mentioned). With this, we would at least have a backlog of items to work on.

Again, I'm fully supportive of your view (and it's what I'm doing at work as well), but with an open source project, we have to take what we can get in some cases.

Please keep pushing for these things - this is great stuff, please keep it coming!

**Comment by cornfeedhobo** [comment]
@nwinkler I'd like to return to this with a longer reply later, but I do think it's worth pointing out that this PR was created by someone that doesn't even use the tool in question. That took extra time to do. This gives me the sense that either the work is motivated out of spite, or that there is plenty of extra capacity to do things right where possible, it just needs the proper motivation.

I completely understand your point about opensource vs a company. I can't stress enough, this only came up because of the context I've laid out. I mean, A few years ago I submitted a pretty bad `GOPATH` parser. Took like a year to fix. No one put up _any_ fight. Yet today we have enough capacity to have these drawn out conversations about a plugin that is doing far less. You have reached critical mass. It's time to add some rigidity for the core contributors. It's either that, or I'm left to assume that the culture is left to devolve into exactly where I see these type of "small code" mantras go: pedantic yak shaving by non-invested parties.

**Comment by nwinkler** [comment]
Thanks - that's a good summary. I'm open for suggestions on how we could make this happen. Would love to hear from both you (@cornfeedhobo) and @davidpfarrell on how could structure Bash-it work in more efficient ways.

A couple of ideas - let me know what you think:

* Moving bigger undertakings (more/better documentation, cleaning up copy/paste work, refactorings) to GitHub Projects instead of handling via individual issues/PRs.
* Setting up a team in GitHub with a separate discussion channel to discuss these things there instead of going through issues/comments.
* We will need a better understanding about people's capacity to work on things. Not sure how we want to track that.

**Comment by cornfeedhobo** [comment]
@nwinkler I'd like to give this some time to chew on. I'm a little stretched thin, but think you're 100% right that this project sits in a vague place where it only needs a light hand. I'd like to get this right because I am passionate about contributing to shared projects like this over maintaining my own one-off solutions. Everyone benefits, plus I can make it an easy standard to push on my teams(!!!). Although MacOS just made that a harder conversation now that they are switching to `zsh` as the default shell. Now I'll have devs writing non-portable scripts more and more, but I digress.

I love these ideas!

> Moving bigger undertakings (more/better documentation, cleaning up copy/paste work, refactorings) to GitHub Projects instead of handling via individual issues/PRs.

Yes we should definitely do this. Everyone benefits.

> Setting up a team in GitHub with a separate discussion channel to discuss these things there instead of going through issues/comments.

My initial thought is to treat it like a small startup; core contributors use email/chat to discuss. I prefer Freenode IRC because I detest email (too easy to come off mean) and it's a common standard for opensource projects. The technical barrier to entry just high enough to discourage the most time-consuming support questions.

> We will need a better understanding about people's capacity to work on things. Not sure how we want to track that.

I think we could use a system of ownership and delegation. E.g. set out projects and tasks. if an issue comes up or someone volunteers, give it an owner. If they fall off the wagon and someone else takes it on, we can ping them for sign-off and review, or re-assign if it they stop replying. That might fit the project for the current state it's in? I've yet to use github projects in a serious capacity and am relating it to pivotal/jira/gus/etc. I'll try to make time to play with it more on a test repo.

I'm definitely going to keep thinking about this when I can. It feels like it's the right time.

Thanks again for starting the project and the relentless positivity!
@davidpfarrell thanks for your passion and support!

**Comment by nwinkler** [comment]
@cornfeedhobo Thanks for the kind words! One (major) clarification: I did not start Bash-it, that was @revans. I started contributing at one point and was then added to the list of owners.

I'll look into setting up some of the mentioned things - if only I had more time...
